File: homeworks/ch5_processes/main.c
/*
 * process exercises: demonstrations of fork, pipe, and i/o redirection
 * questions from ostep chapter 5
 */

#include <stdio.h>
#include <stdlib.h>
#include <sys/fcntl.h>
#include <unistd.h>
#include <string.h>
#include <fcntl.h>
#include <sys/wait.h>

/*
 * q1: fork basics
 * demonstrates how child and parent processes get their own copy of variables
 */
void q1() {
    int x = 100;
    printf("before fork: %d\n", x);
    int rc = fork();

    if (rc < 0) {         // fork failed
        fprintf(stderr, "fork failed\n");
        exit(1);
    } else if (rc == 0) { // child process
        x = 25;
        printf("inside child: %d\n", x);
    } else {              // parent process (rc > 0)
        x = 50;
        printf("inside parent: %d\n", x);
    }

    printf("end: %d\n", x);
}

/*
 * q2: file descriptors and concurrent writing
 * demonstrates how child and parent can share a file descriptor
 * shows concurrent writing to the same file
 */
void q2() {
    // open file BEFORE fork so both processes share same file descriptor
    int fd = open("./q2.output", O_CREAT|O_WRONLY|O_TRUNC, S_IRWXU);

    int rc = fork();
    if (rc < 0) {
        fprintf(stderr, "fork failed\n");
        exit(1);
    } else if (rc == 0) {
        printf("Child has fd: %d\n", fd);
        for (int i = 0; i < 1000; i++) {
            char buf[100];
            sprintf(buf, "child %d\n", i);
            write(fd, buf, strlen(buf));
            // do some work to consume CPU time
            for(int j = 0; j < 1000000; j++) { }
        }
    } else {
        printf("Parent has fd: %d\n", fd);
        for (int i = 0; i < 1000; i++) {
            char buf[100];
            sprintf(buf, "parent %d\n", i);
            write(fd, buf, strlen(buf));
            // Do some work to consume CPU time
            for(int j = 0; j < 1000000; j++) { }
        }
    }

    close(fd);  //both processes close the same fd
}

/*
 * q3: process synchronization without wait
 * uses a pipe to ensure child prints "hello" before parent prints "goodbye"
 */
void q3() {
    int pipefd[2];
    pipe(pipefd);  // create a pipe before forking

    int rc = fork();
    if (rc < 0) {
        fprintf(stderr, "fork failed\n");
        exit(1);
    } else if (rc == 0) {
        printf("hello\n");
        // signal to parent we're done
        write(pipefd[1], "done", 4);
        close(pipefd[1]);
    } else  {
        char buf[10];
        // wait for child's signal
        read(pipefd[0], buf, 4);
        close(pipefd[0]);
        printf("goodbye\n");
    }
}

/*
 * q4: exec variants
 * demonstrates different ways to execute a new program using exec family
 */
void q4() {
    int rc = fork();
    if (rc < 0) {
        fprintf(stderr, "fork failed\n");
        exit(1);
    } else if (rc == 0) {
        // show multiple exec variants
        printf("child: about to exec ls\n");

        // could cycle through these variants:
        execl("/bin/ls", "ls", NULL);
        // execle("/bin/ls", "ls", NULL, (char *[]){NULL});
        // execlp("ls", "ls", NULL);
        // execv("/bin/ls", (char *[]){"ls", NULL});
        // execvp("ls", (char *[]){"ls", NULL});

        printf("this shouldn't print\n");
    } else {
        wait(NULL);
    }
}

/*
 * q5: understanding wait
 * demonstrates how wait() works and what it returns
 */
void q5() {
    int rc = fork();
    if (rc < 0) {
        fprintf(stderr, "fork failed\n");
        exit(1);
    } else if (rc == 0) {
        printf("child: pid %d\n", (int)getpid());
        // try wait in child
        int wc = wait(NULL);
        printf("child wait returned: %d\n", wc);
    } else {
        int wc = wait(NULL);
        printf("parent wait returned: %d\n", wc);
    }
}

/*
 * q6: waitpid versus wait
 * shows how to wait for specific child process
 */
void q6() {
    int rc1 = fork();
    if (rc1 < 0) {
        fprintf(stderr, "fork failed\n");
        exit(1);
    } else if (rc1 == 0) {
        printf("child 1: pid %d\n", (int)getpid());
        sleep(1);
    } else {
        int rc2 = fork();
        if (rc2 < 0) {
            fprintf(stderr, "fork failed\n");
            exit(1);
        } else if (rc2 == 0) {
            printf("child 2: pid %d\n", (int)getpid());
            sleep(2);
        } else {
            waitpid(rc2, NULL, 0);  // wait specifically for second child
            printf("parent: child 2 finished\n");
        }
    }
}

/*
 * q7: closing standard output
 * demonstrates behavior of printf and write after closing stdout
 */
void q7() {
    int rc = fork();

    if (rc < 0) {
        fprintf(stderr, "fork failed\n");
        exit(1);
    } else if (rc == 0) {
        printf("child before closing stdout: this should print\n");
        close(STDOUT_FILENO);  // close stdout

        // try to print after closing stdout
        printf("child after closing stdout: this shouldn't show\n");

        // let's also try write() to see the difference
        write(STDOUT_FILENO, "try direct write\n", 16);

        // can still print to stderr
        fprintf(stderr, "child stderr: this should print\n");
    } else {
        wait(NULL);
        printf("parent: still can print\n");
    }
}

/*
 * q8: pipe between two children
 * implements "ls | wc -l" using fork, pipe, and exec
 * connects stdout of first child to stdin of second child
 */
void q8() {
    int pipefd[2];
    pipe(pipefd);  // create pipe before forking

    int rc1 = fork();  // first fork
    if (rc1 < 0) {
        fprintf(stderr, "fork failed\n");
        exit(1);
    } else if (rc1 == 0) {  // first child (writer)
        close(pipefd[0]);   // close unused read end

        // redirect stdout to pipe write end
        close(STDOUT_FILENO);
        dup2(pipefd[1], STDOUT_FILENO);
        close(pipefd[1]);

        // execute 'ls' ... its output will go to the pipe
        char *args[] = {"ls", NULL};
        execvp(args[0], args);

        fprintf(stderr, "exec failed\n");
        exit(1);
    }

    int rc2 = fork();  // second fork
    if (rc2 < 0) {
        fprintf(stderr, "fork failed\n");
        exit(1);
    } else if (rc2 == 0) {  // second child (reader)
        close(pipefd[1]);   // close unused write end

        // redirect stdin to pipe read end
        close(STDIN_FILENO);
        dup2(pipefd[0], STDIN_FILENO);
        close(pipefd[0]);

        // execute 'wc' ... it will read from the pipe
        char *args[] = {"wc", "-l", NULL};
        execvp(args[0], args);

        fprintf(stderr, "exec failed\n");
        exit(1);
    }

    // parent closes both ends of pipe
    close(pipefd[0]);
    close(pipefd[1]);

    // wait for both children
    waitpid(rc1, NULL, 0);
    waitpid(rc2, NULL, 0);
}

int main() {
    // q1();
    // q2();
    // q3();
    // q4();
    // q5();
    // q6();
    // q7();
    q8();

    return 0;
}

File: homeworks/rust/ch5_processes/src/main.rs
use nix::sys::wait::{wait, waitpid};
use nix::unistd;
use nix::unistd::{fork, pipe, ForkResult};
use std::fs::File;
use std::io::{Read, Write};
use std::os::unix::io::{AsRawFd, FromRawFd};
use std::process::{exit, Command};

#[allow(dead_code)]
mod exercises {
    use super::*;
    // q1: fork basics
    pub fn q1() {
        let x = 100;
        println!("before fork: {}", x);

        match unsafe { fork() } {
            Ok(ForkResult::Parent { child: _ }) => {
                let x = 50;
                println!("inside parent: {}", x);
            }
            Ok(ForkResult::Child) => {
                let x = 25;
                println!("inside child: {}", x);
            }
            Err(_) => {
                eprintln!("fork failed");
                exit(1);
            }
        }
        println!("end: {}", x);
    }

    // q2: file descriptors and concurrent writing
    pub fn q2() {
        let file = File::create("./q2.output").expect("failed to create file");
        let fd = file.as_raw_fd();

        match unsafe { fork() } {
            Ok(ForkResult::Child) => {
                println!("Child has fd: {}", fd);
                for i in 0..1000 {
                    let buf = format!("child {}\n", i);
                    unsafe {
                        let mut file = File::from_raw_fd(fd);
                        file.write_all(buf.as_bytes()).expect("write failed");
                        // don't close file ... avoids closing fd
                        std::mem::forget(file);
                    }
                    // busy work
                    for _ in 0..1000000 {}
                }
            }
            Ok(ForkResult::Parent { child: _ }) => {
                println!("Parent has fd: {}", fd);
                for i in 0..1000 {
                    let buf = format!("parent {}\n", i);
                    unsafe {
                        let mut file = File::from_raw_fd(fd);
                        file.write_all(buf.as_bytes()).expect("write failed");
                        std::mem::forget(file);
                    }
                    // busy work
                    for _ in 0..1000000 {}
                }
            }
            Err(_) => {
                eprintln!("fork failed");
                exit(1);
            }
        }
    }

    // q3: process synchronization without wait
    pub fn q3() {
        let (read_fd, write_fd) = pipe().expect("pipe failed");

        match unsafe { fork() } {
            Ok(ForkResult::Child) => {
                println!("hello");
                // signal to parent we're done
                unsafe {
                    let mut pipe = File::from_raw_fd(write_fd);
                    pipe.write_all(b"done").expect("write failed");
                }
                // read_fd automatically closed
            }
            Ok(ForkResult::Parent { child: _ }) => {
                // Wait for child's signal
                let mut buf = [0u8; 4];
                unsafe {
                    let mut pipe = File::from_raw_fd(read_fd);
                    pipe.read_exact(&mut buf).expect("read failed");
                }
                println!("goodbye");
                // write_fd automatically closed
            }
            Err(_) => {
                eprintln!("fork failed");
                exit(1);
            }
        }
    }

    // q4: exec variants
    pub fn q4() {
        match unsafe { fork() } {
            Ok(ForkResult::Child) => {
                println!("child: about to exec ls");

                // rust equivalent of different exec variants

                // execv style (using Command)
                Command::new("/bin/ls")
                    .spawn()
                    .expect("exec failed")
                    .wait()
                    .expect("wait failed");

                // could also use nix::unistd::execvp directly (more C-like)
                // let args = ["ls".to_string()];
                // nix::unistd::execvp(
                //     &std::ffi::CString::new("ls").unwrap(),
                //     &[&std::ffi::CString::new("ls").unwrap()]
                // ).expect("exec failed");

                println!("this shouldn't print");
                exit(0);
            }
            Ok(ForkResult::Parent { child }) => {
                waitpid(Some(child), None).expect("wait failed");
            }
            Err(_) => {
                eprintln!("fork failed");
                exit(1);
            }
        }
    }

    // q5: understanding wait
    pub fn q5() {
        match unsafe { fork() } {
            Ok(ForkResult::Child) => {
                println!("child: pid {}", unistd::getpid());
                // try wait in child
                match wait() {
                    Ok(status) => println!("child wait returned: {:?}", status),
                    Err(e) => println!("child wait returned error: {}", e),
                }
            }
            Ok(ForkResult::Parent { child }) => match waitpid(Some(child), None) {
                Ok(status) => println!("parent wait returned: {:?}", status),
                Err(e) => println!("parent wait returned error: {}", e),
            },
            Err(_) => {
                eprintln!("fork failed");
                exit(1);
            }
        }
    }

    // q6: waitpid versus wait
    pub fn q6() {
        match unsafe { fork() } {
            Ok(ForkResult::Child) => {
                println!("child 1: pid {}", unistd::getpid());
                std::thread::sleep(std::time::Duration::from_secs(1));
                exit(0);
            }
            Ok(ForkResult::Parent { child: _child1 }) => {
                match unsafe { fork() } {
                    Ok(ForkResult::Child) => {
                        println!("child 2: pid {}", unistd::getpid());
                        std::thread::sleep(std::time::Duration::from_secs(2));
                        exit(0);
                    }
                    Ok(ForkResult::Parent { child: child2 }) => {
                        // wait specifically for second child
                        waitpid(Some(child2), None).expect("waitpid failed");
                        println!("parent: child 2 finished");
                        // could wait for child1 here if needed
                    }
                    Err(_) => {
                        eprintln!("fork failed");
                        exit(1);
                    }
                }
            }
            Err(_) => {
                eprintln!("fork failed");
                exit(1);
            }
        }
    }

    // q7: closing standard output
    pub fn q7() {
        match unsafe { fork() } {
            Ok(ForkResult::Child) => {
                println!("child before closing stdout: this should print");

                // close stdout
                unistd::close(1).expect("close failed");

                // try to print after closing stdout
                println!("child after closing stdout: this shouldn't show");

                // try direct write to stdout
                let message = b"try direct write\n";
                unsafe {
                    let mut stdout = File::from_raw_fd(1);
                    let _ = stdout.write_all(message); // this will fail
                    std::mem::forget(stdout); // don't close fd 1 again
                }

                // can still print to stderr
                eprintln!("child stderr: this should print");
            }
            Ok(ForkResult::Parent { child }) => {
                waitpid(Some(child), None).expect("wait failed");
                println!("parent: still can print");
            }
            Err(_) => {
                eprintln!("fork failed");
                exit(1);
            }
        }
    }

    // q8: pipe between two children
    pub fn q8() {
        let (read_fd, write_fd) = pipe().expect("pipe failed");

        match unsafe { fork() } {
            Ok(ForkResult::Child) => {
                // first child (writer)
                nix::unistd::close(read_fd).expect("close failed");
                nix::unistd::dup2(write_fd, 1).expect("dup2 failed");
                nix::unistd::close(write_fd).expect("close failed");

                Command::new("ls")
                    .spawn()
                    .expect("ls failed")
                    .wait()
                    .expect("wait failed");
                exit(0);
            }
            Ok(ForkResult::Parent { child: child1 }) => {
                match unsafe { fork() } {
                    Ok(ForkResult::Child) => {
                        // second child (reader)
                        nix::unistd::close(write_fd).expect("close failed");
                        nix::unistd::dup2(read_fd, 0).expect("dup2 failed");
                        nix::unistd::close(read_fd).expect("close failed");

                        Command::new("wc")
                            .arg("-l")
                            .spawn()
                            .expect("wc failed")
                            .wait()
                            .expect("wait failed");
                        exit(0);
                    }
                    Ok(ForkResult::Parent { child: child2 }) => {
                        // parent
                        nix::unistd::close(read_fd).expect("close failed");
                        nix::unistd::close(write_fd).expect("close failed");

                        waitpid(Some(child1), None).expect("waitpid failed");
                        waitpid(Some(child2), None).expect("waitpid failed");
                    }
                    Err(_) => {
                        eprintln!("fork failed");
                        exit(1);
                    }
                }
            }
            Err(_) => {
                eprintln!("fork failed");
                exit(1);
            }
        }
    }
}

fn main() {
    // exercises::q1();
    // exercises::q2();
    // exercises::q3();
    // exercises::q4();
    // exercises::q5();
    // exercises::q6();
    // exercises::q7();
    exercises::q8();
}
